abci/src/codec.rs

```diff
@@ -0,0 +1,174 @@
+            write_buf: BytesMut::new(),
+            _incoming: Default::default(),
+            _outgoing: Default::default(),
+        }
+    }
+}
+
+// Iterating over a codec produces instances of `Result<I>`.
+impl<S, I, O> Iterator for Codec<S, I, O>
+where
+    S: Read,
+    I: Message + Default,
+{
+    type Item = Result<I>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            // Try to decode an incoming message from our buffer first
+            match decode_length_delimited::<I>(&mut self.read_buf) {
+                Ok(opt) => {
+                    if let Some(incoming) = opt {
+                        return Some(Ok(incoming));
+                    }
+                }
+                Err(e) => return Some(Err(e)),
+            }
+
+            // If we don't have enough data to decode a message, try to read
+            // more
+            let bytes_read = match self.stream.read(self.read_window.as_mut()) {
+                Ok(br) => br,
+                Err(e) => return Some(Err(e.into())),
+            };
+            if bytes_read == 0 {
+                // The underlying stream terminated
+                return None;
+            }
+            self.read_buf
+                .extend_from_slice(&self.read_window[..bytes_read]);
+        }
+    }
+}
+
+impl<S, I, O> Codec<S, I, O>
+where
+    S: Write,
+    O: Message,
+{
+    /// Send a message using this codec.
+    pub fn send(&mut self, message: O) -> Result<()> {
+        encode_length_delimited(message, &mut self.write_buf)?;
+        while !self.write_buf.is_empty() {
+            let bytes_written = self.stream.write(self.write_buf.as_ref())?;
+            if bytes_written == 0 {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::WriteZero,
+                    "failed to write to underlying stream",
+                )
+                .into());
+            }
+            self.write_buf.advance(bytes_written);
+        }
+        Ok(self.stream.flush()?)
+    }
+}
+
+/// Encode the given message with a length prefix.
+pub fn encode_length_delimited<M, B>(message: M, mut dst: &mut B) -> Result<()>
+where
+    M: Message,
+    B: BufMut,
+{
+    let mut buf = BytesMut::new();
+    message.encode(&mut buf)?;
+    let buf = buf.freeze();
+    encode_varint(buf.len() as u64, &mut dst);
+    dst.put(buf);
+    Ok(())
+}
+
+/// Attempt to decode a message of type `M` from the given source buffer.
+pub fn decode_length_delimited<M>(src: &mut BytesMut) -> Result<Option<M>>
+where
+    M: Message + Default,
+{
+    let src_len = src.len();
+    let mut tmp = src.clone().freeze();
+    let encoded_len = match decode_varint(&mut tmp) {
+        Ok(len) => len,
+        Err(e) => {
+            return if src_len <= MAX_VARINT_LENGTH {
+                // We've potentially only received a partial length delimiter
+                Ok(None)
+            } else {
+                Err(e)
+            };
+        }
```

Comment:
The following reads a bit better to me, but again feel free to ignore.

```suggestion
        // We've potentially only received a partial length delimiter
        Err(e) if src_len <= MAX_VARINT_LENGTH => return Ok(None),
        Err(e) => return Err(e),
```

Reply:
Applied :+1: